src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs

```diff
@@ -0,0 +1,822 @@
+        }
+
+        private static void WriteCallback(IAsyncResult transportResult)
+        {
+            if ((transportResult.AsyncState is LazyAsyncResult))
+            {
+                if (GlobalLog.IsEnabled)
+                {
+                    GlobalLog.Assert("WriteCallback|State type is wrong, expected LazyAsyncResult.");
+                }
+                Debug.Fail("WriteCallback|State type is wrong, expected LazyAsyncResult.");
+            }
+
+            if (transportResult.CompletedSynchronously)
+            {
+                return;
+            }
+
+            LazyAsyncResult lazyResult = (LazyAsyncResult)transportResult.AsyncState;
+
+            // Async completion.
+            try
+            {
+                NegoState authState = (NegoState)lazyResult.AsyncObject;
+                authState._framer.EndWriteMessage(transportResult);
+
+                // Special case for an error notification.
+                if (lazyResult.Result is Exception)
+                {
+                    authState._canRetryAuthentication = true;
+                    throw (Exception)lazyResult.Result;
+                }
+
+                authState.CheckCompletionBeforeNextReceive(lazyResult);
+            }
+            catch (Exception e)
+            {
+                if (lazyResult.InternalPeekCompleted)
+                {
+                    // This will throw on a worker thread.
+                    throw;
+                }
+
+                lazyResult.InvokeCallback(e);
+            }
+        }
+
+        private static void ReadCallback(IAsyncResult transportResult)
+        {
+            if ((transportResult.AsyncState is LazyAsyncResult))
+            {
+                if (GlobalLog.IsEnabled)
+                {
+                    GlobalLog.Assert("ReadCallback|State type is wrong, expected LazyAsyncResult.");
+                }
+                Debug.Fail("ReadCallback|State type is wrong, expected LazyAsyncResult.");
+            }
+
+            if (transportResult.CompletedSynchronously)
+            {
+                return;
+            }
+
+            LazyAsyncResult lazyResult = (LazyAsyncResult)transportResult.AsyncState;
+
+            // Async completion.
+            try
+            {
+                NegoState authState = (NegoState)lazyResult.AsyncObject;
+                byte[] message = authState._framer.EndReadMessage(transportResult);
+                authState.ProcessReceivedBlob(message, lazyResult);
+            }
+            catch (Exception e)
+            {
+                if (lazyResult.InternalPeekCompleted)
+                {
+                    // This will throw on a worker thread.
+                    throw;
+                }
+
+                lazyResult.InvokeCallback(e);
+            }
+        }
+
+        internal static bool IsError(SecurityStatusPal status)
+        {
+            return ((int)status >= (int)SecurityStatusPal.OutOfMemory);
```

Comment:
Okay, I guess that it does have a comment indicating that OutOfMemory and beyond are "Errors", but since there's no gap in that enumeration I'm concerned about future expansion needs.

Since this code is assuming that everything with a higher value than SecurityStatusPal.OutOfMemory is an error, there really should be a gap, so possibly assign it equal to 0x010000 (or whatever) to leave room for later.

The casts seem completely unnecessary to me... enums are comparable with `>=`, so this line should be able to be `return (status >= SecurityStatusCode.OutOfMemory);`.